### DISCUSSION TITLE
Extended APIRouter class don't validate the data

### DISCUSSION THREAD

**Discussion by Nuli18** [fastapi/fastapi#9964] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import TypeVar
from pydantic import BaseModel
from fastapi import APIRouter, FastAPI

# Data model for validation
class ManufacturerCreate(BaseModel):
    name: str


# Generic for inheritance
CreateType = TypeVar("CreateType", bound=BaseModel)

# Base router for generics CRUD operations
class BaseRouter(APIRouter):
    def __init__(self, create_model: CreateType):
        super().__init__()
        self.create_model = create_model
        self.setup_routes()

    def setup_routes(self):
        self.add_api_route(
            "/", self.create, methods=["POST"], response_model=self.create_model)

    async def create(self, data: CreateType):
        return data

# Data router for customize validation for data type
class ManufacturerRouter(BaseRouter):
    def __init__(self):
        super().__init__(create_model=ManufacturerCreate)


app = FastAPI()

app.include_router(ManufacturerRouter())
```

### Description

I tried to make a base router for my generic CRUD operations and make it inherited by datatype router. But when i try to request the path with data it gives me `422 Unprocessable Entity`. 

Data in question : 
```
{
    "name": "APPLE"
}
```

Response : 
```
{
    "detail": [
        {
            "type": "missing",
            "loc": [
                "query",
                "data"
            ],
            "msg": "Field required",
            "input": null,
            "url": "https://errors.pydantic.dev/2.1/v/missing"
        }
    ]
}
```

Am i missing something ?

### Operating System

Linux, Windows

### Operating System Details

 wsl.exe --version
Version WSL : 1.2.5.0
Version du noyau : 5.15.90.1
Version WSLg : 1.0.51
Version MSRDC : 1.2.3770
Version direct3D : 1.608.2-61064218
Version de DXCore : 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Version de Windows : 10.0.19045.2965

 cat /proc/version
Linux version 5.15.90.1-microsoft-standard-WSL2 (oe-user@oe-host) (x86_64-msft-linux-gcc (GCC) 9.3.0, GNU ld (GNU Binutils) 2.34.0.20200220) #1 SMP Fri Jan 27 02:56:13 UTC 2023

 cat /etc/*-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/

### FastAPI Version

0.100.0

### Python Version

Python 3.9.2

**Comment by arjwilliams** [comment, marked as answer]
Hi @Nuli18 ,

The issue is that the type of `data` in the `create` method is being recognised as `CreateType`, rather than `ManufacturerCreate`, because you don't change the type annotation on the `create` method, and so when you make the request, FastAPI doesn't recognise the incoming data.

You will need to wrap the `create` method in order to have dynamic argument types, so that FastAPI will be able to recognise the incoming data. If you want to do this over multiple subclasses, it might be cleaner to use a metaclass for this:

```python
# Base router for generics CRUD operations
class RouterBase(APIRouter):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.setup_routes()

    def setup_routes(self):
        self.add_api_route("/", self.create, methods=["POST"])

    async def create(self, data: CreateType) -> CreateType:
        # Default implementation
        return data

class MetaRouter(type):
    def __new__(cls, name, bases, attrs, CreateType):
        # So we don't need to explicitly specify RouterBase as a base class of ManufacturerRouter
        if RouterBase not in bases:
            bases = (RouterBase,) + bases

        # Define the create method, which just calls the default RouterBase create method
        # This uses the passed in CreateType argument, rather than the global one,
        #     so specifies both dynamic argument and return type
        async def create(self, data: CreateType) -> CreateType:
            return await super(self.__class__, self).create(data)

        # Assign the create method to the new class
        attrs.setdefault(create.__name__, create)

        return super().__new__(cls, name, bases, attrs)

# Data router for customize validation for data type
class ManufacturerRouter(metaclass=MetaRouter, CreateType=ManufacturerCreate):
    pass
```

If you ever need to override the default implementation, you can still do that in the subclass:

```python
class ManufacturerRouter(metaclass=MetaRouter, CreateType=ManufacturerCreate):
    async def create(self, data: ManufacturerCreate) -> ManufacturerCreate:
       return {"name": "TEST"}
```
